# Problem Statement

When the company receives a loan application, the company has to make a decision for loan approval based on the applicant’s profile. Two types of risks are associated with the bank’s decision:

- If the applicant is likely to repay the loan, then not approving the loan results in a loss of business to the company

- If the applicant is not likely to repay the loan, i.e. he/she is likely to default, then approving the loan may lead to a financial loss for the company

The data provided contains the information about past loan applicants and whether they ‘defaulted’ or not. The aim is to identify patterns which indicate if a person is likely to default, which may be used for taking actions such as denying the loan, reducing the amount of loan, lending (to risky applicants) at a higher interest rate, etc.

Use EDA to understand how consumer attributes and loan attributes influence the tendency of default.

# Data Understanding

In this section, we will try to get an initial understanding of the data provided.

In [1]:
#Load the necessary Libraries
import pandas as pd
import numpy as np
import warnings

In [2]:
# Set required options 
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 200)

In [3]:
# Load the loan data from CSV
loan_data = pd.read_csv('loan.csv')
loan_data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-11   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-11  Charged Off   
2           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65            0   
1                  bike    309xx         GA   1.00            0   
2  real estate business    606xx         IL   8.72            0   
3              personel    917xx         CA  20.00            0   
4              Personal    972xx         OR  17.94            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85               1                     NaN   
1           Apr-99               5                     NaN   
2           Nov-01               2                     NaN   
3           Feb-96               1                    35.0   
4           Jan-96               0                    38.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         3        0      13648     83.70%          9   
1                     NaN         3        0       1687      9.40%          4   
2                     NaN         2        0       2956     98.50%         10   
3                     NaN        10        0       5598        21%         37   
4                     NaN        15        0      27783     53.90%         38   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   f       0.00           0.00   5863.155187   
1                   f       0.00           0.00   1008.710000   
2                   f       0.00           0.00   3005.666844   
3                   f       0.00           0.00  12231.890000   
4                   f     524.06         524.06   3513.330000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5833.84          5000.00         863.16     

In [4]:
# Load Data Dictionary
loan_data_dictionary = pd.read_excel('Data_Dictionary.xlsx')
loan_data_dictionary

LoanStatNew  \
0                    acc_now_delinq   
1              acc_open_past_24mths   
2                        addr_state   
3                          all_util   
4                        annual_inc   
5                  annual_inc_joint   
6                  application_type   
7                       avg_cur_bal   
8                    bc_open_to_buy   
9                           bc_util   
10         chargeoff_within_12_mths   
11          collection_recovery_fee   
12       collections_12_mths_ex_med   
13                      delinq_2yrs   
14                      delinq_amnt   
15                             desc   
16                              dti   
17                        dti_joint   
18                 earliest_cr_line   
19                       emp_length   
20                        emp_title   
21                  fico_range_high   
22                   fico_range_low   
23                      funded_amnt   
24                  funded_amnt_inv   
25                            grade   
26                   home_ownership   
27                               id   
28                          il_util   
29              initial_list_status   
30                           inq_fi   
31                     inq_last_12m   
32                   inq_last_6mths   
33                      installment   
34                         int_rate   
35                          issue_d   
36               last_credit_pull_d   
37             last_fico_range_high   
38              last_fico_range_low   
39                  last_pymnt_amnt   
40                     last_pymnt_d   
41                        loan_amnt   
42                      loan_status   
43                       max_bal_bc   
44                        member_id   
45               mo_sin_old_il_acct   
46             mo_sin_old_rev_tl_op   
47            mo_sin_rcnt_rev_tl_op   
48                   mo_sin_rcnt_tl   
49                         mort_acc   
50           mths_since_last_delinq   
51      mths_since_last_major_derog   
52           mths_since_last_record   
53               mths_since_rcnt_il   
54             mths_since_recent_bc   
55         mths_since_recent_bc_dlq   
56            mths_since_recent_inq   
57   mths_since_recent_revol_delinq   
58                     next_pymnt_d   
59            num_accts_ever_120_pd   
60                   num_actv_bc_tl   
61                  num_actv_rev_tl   
62                      num_bc_sats   
63                        num_bc_tl   
64                        num_il_tl   
65                    num_op_rev_tl   
66                    num_rev_accts   
67              num_rev_tl_bal_gt_0   
68                         num_sats   
69                 num_tl_120dpd_2m   
70                     num_tl_30dpd   
71               num_tl_90g_dpd_24m   
72               num_tl_op_past_12m   
73                         open_acc   
74                      open_acc_6m   
75                      open_il_12m   
76                      open_il_24m   
77                       open_il_6m   
78                      open_rv_12m   
79                      open_rv_24m   
80                        out_prncp   
81                    out_prncp_inv   
82                   pct_tl_nvr_dlq   
83                 percent_bc_gt_75   
84                      policy_code   
85                          pub_rec   
86             pub_rec_bankruptcies   
87                          purpose   
88                       pymnt_plan   
89                       recoveries   
90                        revol_bal   
91                       revol_util   
92                        sub_grade   
93                        tax_liens   
94                             term   
95                            title   
96                     tot_coll_amt   
97                      tot_cur_bal   
98                  tot_hi_cred_lim   
99                        total_acc   
100               total_bal_ex_mort   
101                    total_bal_il   
102    

## Dataframe Summary

Check details of the dataframe like how many rows and how many columns are present.

In [5]:
loan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(74), int64(13), object(24)
memory usage: 33.6+ MB


As per above the dataframe have 39717 rows with 111 columns each.

In [6]:
loan_data.describe()

id     member_id     loan_amnt   funded_amnt  \
count  3.971700e+04  3.971700e+04  39717.000000  39717.000000   
mean   6.831319e+05  8.504636e+05  11219.443815  10947.713196   
std    2.106941e+05  2.656783e+05   7456.670694   7187.238670   
min    5.473400e+04  7.069900e+04    500.000000    500.000000   
25%    5.162210e+05  6.667800e+05   5500.000000   5400.000000   
50%    6.656650e+05  8.508120e+05  10000.000000   9600.000000   
75%    8.377550e+05  1.047339e+06  15000.000000  15000.000000   
max    1.077501e+06  1.314167e+06  35000.000000  35000.000000   

       funded_amnt_inv   installment    annual_inc           dti  \
count     39717.000000  39717.000000  3.971700e+04  39717.000000   
mean      10397.448868    324.561922  6.896893e+04     13.315130   
std        7128.450439    208.874874  6.379377e+04      6.678594   
min           0.000000     15.690000  4.000000e+03      0.000000   
25%        5000.000000    167.020000  4.040400e+04      8.170000   
50%        8975.000000    280.220000  5.900000e+04     13.400000   
75%       14400.000000    430.780000  8.230000e+04     18.600000   
max       35000.000000   1305.190000  6.000000e+06     29.990000   

        delinq_2yrs  inq_last_6mths  mths_since_last_delinq  \
count  39717.000000    39717.000000            14035.000000   
mean       0.146512        0.869200               35.900962   
std        0.491812        1.070219               22.020060   
min        0.000000        0.000000                0.000000   
25%        0.000000        0.000000               18.000000   
50%        0.000000        1.000000               34.000000   
75%        0.000000        1.000000               52.000000   
max       11.000000        8.000000              120.000000   

       mths_since_last_record      open_acc       pub_rec      revol_bal  \
count             2786.000000  39717.000000  39717.000000   39717.000000   
mean                69.698134      9.294408      0.055065   13382.528086   
std                 43.822529      4.400282      0.237200   15885.016641   
min                  0.000000      2.000000      0.000000       0.000000   
25%                 22.000000      6.000000      0.000000    3703.000000   
50%                 90.000000      9.000000      0.000000    8850.000000   
75%                104.000000     12.000000      0.000000   17058.000000   
max                129.000000     44.000000      4.000000  149588.000000   

          total_acc     out_prncp  out_prncp_inv   total_pymnt  \
count  39717.000000  39717.000000   39717.000000  39717.000000   
mean      22.088828     51.227887      50.989768  12153.596544   
std       11.401709    375.172839     373.824457   9042.040766   
min        2.000000      0.000000       0.000000      0.000000   
25%       13.000000      0.000000       0.000000   5576.930000   
50%       20.000000      0.000000       0.000000   9899.640319   
75%       29.000000      0.000000       0.000000  16534.433040   
max       90.000000   6311.470000    6307.370000  58563.679930   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
count     39717.000000     39717.000000   39717.000000        39717.000000   
mean      11567.149118      9793.348813    2263.663172            1.363015   
std        8942.672613      7065.522127    2608.111964            7.289979   
min           0.000000         0.000000       0.000000            0.000000   
25%        5112.310000      4600.000000     662.180000            0.000000   
50%        9287.150000      8000.000000    1348.910000            0.000000   
75%       15798.810000     13653.260000    2833.400000            0.000000   
max       58563.680000     35000.020000   23563.680000          180.200000   

         recoveries  collection_recovery_fee  last_pymnt_amnt  \
count  39717.000000             39717.000000     39717.000000   
mean      95.221624                12.406112      2678.826162   
std      688.744771               148.671593      4447.136012   
min    

The summary metrics of all columns are shown above.

### Column Types

The type of each column in the original dataframe is shown below.

In [7]:
loan_data.dtypes

id                                  int64
member_id                           int64
loan_amnt                           int64
funded_amnt                         int64
funded_amnt_inv                   float64
term                               object
int_rate                           object
installment                       float64
grade                              object
sub_grade                          object
emp_title                          object
emp_length                         object
home_ownership                     object
annual_inc                        float64
verification_status                object
issue_d                            object
loan_status                        object
pymnt_plan                         object
url                                object
desc                               object
purpose                            object
title                              object
zip_code                           object
addr_state                        

## Check Null Value Counts

Analyse the null values in the dataframe

In [8]:
# Checking all the columns with null values
null_count = loan_data.isnull().sum()
null_count[null_count > 0].sort_values(ascending=False)

bc_open_to_buy                    39717
acc_open_past_24mths              39717
bc_util                           39717
mo_sin_old_il_acct                39717
mo_sin_old_rev_tl_op              39717
mo_sin_rcnt_rev_tl_op             39717
mo_sin_rcnt_tl                    39717
mort_acc                          39717
mths_since_recent_bc              39717
mths_since_recent_bc_dlq          39717
mths_since_recent_inq             39717
mths_since_recent_revol_delinq    39717
num_accts_ever_120_pd             39717
num_actv_bc_tl                    39717
num_actv_rev_tl                   39717
num_bc_sats                       39717
num_bc_tl                         39717
num_il_tl                         39717
num_op_rev_tl                     39717
num_rev_accts                     39717
num_rev_tl_bal_gt_0               39717
num_sats                          39717
num_tl_120dpd_2m                  39717
num_tl_30dpd                      39717
num_tl_90g_dpd_24m                39717


# Data Cleanup

In this section we will clean up the data.

## Missing Value Handline

Drop columns and rows which are having significant number of missing values.

### Drop Null Columns
We have seen above that few of the rows have significant number of null values. We can drop these columns as these won't add any meaningful information to our analysis. We will consider the columns having only less than 60% values as null.

Columns having more than 60% rows having null values are below

In [9]:
missing_values = round((loan_data.isnull().sum()/len(loan_data.id))*100, 2)
missing_values_60_percent = missing_values[missing_values>60].sort_values(ascending=False)
missing_values_60_percent

mo_sin_old_il_acct                100.00
mo_sin_old_rev_tl_op              100.00
mo_sin_rcnt_tl                    100.00
mort_acc                          100.00
mths_since_recent_bc              100.00
mths_since_recent_bc_dlq          100.00
mths_since_recent_inq             100.00
mths_since_recent_revol_delinq    100.00
num_accts_ever_120_pd             100.00
num_actv_bc_tl                    100.00
num_actv_rev_tl                   100.00
num_bc_sats                       100.00
num_bc_tl                         100.00
num_il_tl                         100.00
num_op_rev_tl                     100.00
num_rev_accts                     100.00
num_rev_tl_bal_gt_0               100.00
num_sats                          100.00
num_tl_120dpd_2m                  100.00
num_tl_30dpd                      100.00
num_tl_90g_dpd_24m                100.00
num_tl_op_past_12m                100.00
pct_tl_nvr_dlq                    100.00
percent_bc_gt_75                  100.00
tot_hi_cred_lim 

We can safely drop the above columns as majority of the values are null and hence won't give any useful insights during analysis.

In [10]:
loan_data = loan_data.drop(missing_values_60_percent.index, axis=1)
print('Dropped %d columns' % len(missing_values_60_percent.index.to_list()))

Dropped 57 columns


Verify that the missing values in columns is less than 60%

In [11]:
missing_values = round((loan_data.isnull().sum()/len(loan_data.id))*100, 2)
missing_values = pd.DataFrame(missing_values[missing_values>0], columns=["missing_value_percentage"])
missing_values.sort_values(by=["missing_value_percentage"], ascending=False)

missing_value_percentage
desc                                           32.58
emp_title                                       6.19
emp_length                                      2.71
pub_rec_bankruptcies                            1.75
last_pymnt_d                                    0.18
collections_12_mths_ex_med                      0.14
chargeoff_within_12_mths                        0.14
revol_util                                      0.13
tax_liens                                       0.10
title                                           0.03
last_credit_pull_d                              0.01

### Drop Columns Having Same Value for All Rows

Some of the columns would be having same value for all the rows and hence not useful for analysis. These columns can be dropped.

In [12]:
columns_with_unique_value_counts = loan_data.nunique().sort_values(ascending=True)
columns_with_unique_value_counts

tax_liens                         1
delinq_amnt                       1
chargeoff_within_12_mths          1
acc_now_delinq                    1
application_type                  1
policy_code                       1
initial_list_status               1
collections_12_mths_ex_med        1
pymnt_plan                        1
term                              2
pub_rec_bankruptcies              3
verification_status               3
loan_status                       3
pub_rec                           5
home_ownership                    5
grade                             7
inq_last_6mths                    9
emp_length                       11
delinq_2yrs                      11
purpose                          14
sub_grade                        35
open_acc                         40
addr_state                       50
issue_d                          55
total_acc                        82
last_pymnt_d                    101
last_credit_pull_d              106
int_rate                    

Any columns with only one value can be dropped.

Below check shows the columns with single values.

In [13]:
# Finding columns with unique values
columns_with_single_value = columns_with_unique_value_counts[columns_with_unique_value_counts == 1]
print(columns_with_single_value)

tax_liens                     1
delinq_amnt                   1
chargeoff_within_12_mths      1
acc_now_delinq                1
application_type              1
policy_code                   1
initial_list_status           1
collections_12_mths_ex_med    1
pymnt_plan                    1
dtype: int64


The above columns can be dropped from the dataframe as they are having a single value for all rows

In [14]:
# Dropping columns with unique values
loan_data = loan_data.drop(columns=columns_with_single_value.index)
loan_data

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0      1077501    1296599       5000         5000           4975.0   
1      1077430    1314167       2500         2500           2500.0   
2      1077175    1313524       2400         2400           2400.0   
3      1076863    1277178      10000        10000          10000.0   
4      1075358    1311748       3000         3000           3000.0   
...        ...        ...        ...          ...              ...   
39712    92187      92174       2500         2500           1075.0   
39713    90665      90607       8500         8500            875.0   
39714    90395      90390       5000         5000           1325.0   
39715    90376      89243       5000         5000            650.0   
39716    87023      86999       7500         7500            800.0   

             term int_rate  installment grade sub_grade  \
0       36 months   10.65%       162.87     B        B2   
1       60 months   15.27%        59.83     C        C4   
2       36 months   15.96%        84.33     C        C5   
3       36 months   13.49%       339.31     C        C1   
4       60 months   12.69%        67.79     B        B5   
...           ...      ...          ...   ...       ...   
39712   36 months    8.07%        78.42     A        A4   
39713   36 months   10.28%       275.38     C        C1   
39714   36 months    8.07%       156.84     A        A4   
39715   36 months    7.43%       155.38     A        A2   
39716   36 months   13.75%       255.43     E        E2   

                        emp_title emp_length home_ownership  annual_inc  \
0                             NaN  10+ years           RENT     24000.0   
1                           Ryder   < 1 year           RENT     30000.0   
2                             NaN  10+ years           RENT     12252.0   
3             AIR RESOURCES BOARD  10+ years           RENT     49200.0   
4        University Medical Group     1 year           RENT     80000.0   
...                           ...        ...            ...         ...   
39712             FiSite Research    4 years       MORTGAGE    110000.0   
39713  Squarewave Solutions, Ltd.    3 years           RENT     18000.0   
39714                         NaN   < 1 year       MORTGAGE    100000.0   
39715                         NaN   < 1 year       MORTGAGE    200000.0   
39716            Evergreen Center   < 1 year            OWN     22000.0   

      verification_status issue_d  loan_status  \
0                Verified  Dec-11   Fully Paid   
1         Source Verified  Dec-11  Charged Off   
2            Not Verified  Dec-11   Fully Paid   
3         Source Verified  Dec-11   Fully Paid   
4         Source Verified  Dec-11      Current   
...                   ...     ...          ...   
39712        Not Verified  Jul-07   Fully Paid   
39713        Not Verified  Jul-07   Fully Paid   
39714        Not Verified  Jul-07   Fully Paid   
39715        Not Verified  Jul-07   Fully Paid   
39716        Not Verified  Jun-07   Fully Paid   

                                                     url  \
0      https://lendingclub.com/browse/loanDetail.acti...   
1      https://lendingclub.com/browse/loanDetail.acti...   
2      https://lendingclub.com/browse/loanDetail.acti...   
3      https://lendingclub.com/browse/loanDetail.acti...   
4      https://lendingclub.com/browse/loanDetail.acti...   
...                                                  ...   
39712  https://lendingclub.com/browse/loanDetail.acti...   
39713  https://lendingclub.com/browse/loanDetail.acti...   
39714  https://lendingclub.com/browse/loanDetail.acti...   
39715  https://lendingclub.com/browse/loanDetail.acti...   
39716  https://lendingclub.com/browse/loanDetail.acti...   

                                                    desc             purpose  \
0        Borrower added on 12/22/11 > I need to upgra...         credit_card   
1        Borrower added on 12/22/11 > I plan to use t...                 car   

### Drop Null Rows

If any rows are having all values as null, we can drop those columns

In [15]:
# Checking how many rows are having all columns as null
loan_data.isnull().all(axis=1).sum()

0

We don't have any rows having all columns as null. So no need to drop any rows as part of this check.

## Standardising Values

Some fields would be having invalid values, which would make it difficult to analyze. We will clean this up in this section.

We can start by checking the type of each column

In [16]:
loan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 45 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       39717 non-null  int64  
 1   member_id                39717 non-null  int64  
 2   loan_amnt                39717 non-null  int64  
 3   funded_amnt              39717 non-null  int64  
 4   funded_amnt_inv          39717 non-null  float64
 5   term                     39717 non-null  object 
 6   int_rate                 39717 non-null  object 
 7   installment              39717 non-null  float64
 8   grade                    39717 non-null  object 
 9   sub_grade                39717 non-null  object 
 10  emp_title                37258 non-null  object 
 11  emp_length               38642 non-null  object 
 12  home_ownership           39717 non-null  object 
 13  annual_inc               39717 non-null  float64
 14  verification_status   

### Column - term

In [17]:
loan_data['term'].value_counts()

 36 months    29096
 60 months    10621
Name: term, dtype: int64

As seen in the value counts, terms is number of months. We can convert the value to integers. The column can also be renamed to indicate that it is number of months.

In [18]:
loan_data['term'] = loan_data['term'].apply(lambda x: int(str(x).replace(' months', '')))
loan_data = loan_data.rename(columns={'term': 'term_in_months'})
loan_data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0  1077501    1296599       5000         5000           4975.0   
1  1077430    1314167       2500         2500           2500.0   
2  1077175    1313524       2400         2400           2400.0   
3  1076863    1277178      10000        10000          10000.0   
4  1075358    1311748       3000         3000           3000.0   

   term_in_months int_rate  installment grade sub_grade  \
0              36   10.65%       162.87     B        B2   
1              60   15.27%        59.83     C        C4   
2              36   15.96%        84.33     C        C5   
3              36   13.49%       339.31     C        C1   
4              60   12.69%        67.79     B        B5   

                  emp_title emp_length home_ownership  annual_inc  \
0                       NaN  10+ years           RENT     24000.0   
1                     Ryder   < 1 year           RENT     30000.0   
2                       NaN  10+ years           RENT     12252.0   
3       AIR RESOURCES BOARD  10+ years           RENT     49200.0   
4  University Medical Group     1 year           RENT     80000.0   

  verification_status issue_d  loan_status  \
0            Verified  Dec-11   Fully Paid   
1     Source Verified  Dec-11  Charged Off   
2        Not Verified  Dec-11   Fully Paid   
3     Source Verified  Dec-11   Fully Paid   
4     Source Verified  Dec-11      Current   

                                                 url  \
0  https://lendingclub.com/browse/loanDetail.acti...   
1  https://lendingclub.com/browse/loanDetail.acti...   
2  https://lendingclub.com/browse/loanDetail.acti...   
3  https://lendingclub.com/browse/loanDetail.acti...   
4  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65            0   
1                  bike    309xx         GA   1.00            0   
2  real estate business    606xx         IL   8.72            0   
3              personel    917xx         CA  20.00            0   
4              Personal    972xx         OR  17.94            0   

  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal revol_util  \
0           Jan-85               1         3        0      13648     83.70%   
1           Apr-99               5         3        0       1687      9.40%   
2           Nov-01               2         2        0       2956     98.50%   
3           Feb-96               1        10        0       5598        21%   
4           Jan-96               0        15        0      27783     53.90%   

   total_acc  out_prncp  out_prncp_inv   total_pymnt  total_pymnt_inv  \
0          9       0.00           0.00   5863.155187          5833.84   
1          4       0.00           0.00   1008.710000          1008.71   
2         10       0.00           0.00   3005.666844          3005.67   
3         37       0.00           0.00  12231.890000         12231.89   
4         38     524.06         524.06   3513.330000          3513.33   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5000.00         863.16                0.00        0.00   
1           456.46         435.17                0.00      117.08   
2          2400.00         605.67                0.00        0.00   
3         10000.00        2214.92               16.97        0.00   
4          2475.94        1037.39                0.00        0.00   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt last_

### Column - int_rate

Interest rate is a percentage value, but since the percentage symbol is present, its stored as an object. We can remove the percentage and convert to float.

In [19]:
loan_data['int_rate'] = loan_data['int_rate'].apply(lambda x: float(str(x).replace('%', '')))

### Column - emp_length

This column indicates how many years of work experience the borrower have. We will start by checking the distinct values.

In [20]:
emp_length_counts = loan_data['emp_length'].value_counts(dropna=False)
emp_length_counts

10+ years    8879
< 1 year     4583
2 years      4388
3 years      4095
4 years      3436
5 years      3282
1 year       3240
6 years      2229
7 years      1773
8 years      1479
9 years      1258
NaN          1075
Name: emp_length, dtype: int64

#### Handle NaN
We can see that there are few NaN values for few rows. These have to be corrected. This could be because the the employment information is not available, or the person have no experience. We cannot confirm on this and we can drop the NaN rows if the number of rows dropped is not too high. 

First we will check the percentage of rows having NaN.

In [21]:
round(emp_length_counts[np.NaN] * 100 / len(loan_data['emp_length']), 2)

2.71

Since the number of rows are only 2.71% we can drop it and pick only the accurate rows for analysis.

In [22]:
loan_data = loan_data[~loan_data['emp_length'].isna()]
loan_data.shape

(38642, 45)

#### Convert Experiences to Int

We can convert the employment length to numeric value, representing the number of years. For 10+ we will provide the value as 10 and for <1 we will provide the value as 0.

In [23]:
loan_data['emp_length'] = loan_data['emp_length'].apply(lambda x: int(str(x).replace(' years', '').replace(' year', '').replace('< 1', '0').replace('10+', '10')))
loan_data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0  1077501    1296599       5000         5000           4975.0   
1  1077430    1314167       2500         2500           2500.0   
2  1077175    1313524       2400         2400           2400.0   
3  1076863    1277178      10000        10000          10000.0   
4  1075358    1311748       3000         3000           3000.0   

   term_in_months  int_rate  installment grade sub_grade  \
0              36     10.65       162.87     B        B2   
1              60     15.27        59.83     C        C4   
2              36     15.96        84.33     C        C5   
3              36     13.49       339.31     C        C1   
4              60     12.69        67.79     B        B5   

                  emp_title  emp_length home_ownership  annual_inc  \
0                       NaN          10           RENT     24000.0   
1                     Ryder           0           RENT     30000.0   
2                       NaN          10           RENT     12252.0   
3       AIR RESOURCES BOARD          10           RENT     49200.0   
4  University Medical Group           1           RENT     80000.0   

  verification_status issue_d  loan_status  \
0            Verified  Dec-11   Fully Paid   
1     Source Verified  Dec-11  Charged Off   
2        Not Verified  Dec-11   Fully Paid   
3     Source Verified  Dec-11   Fully Paid   
4     Source Verified  Dec-11      Current   

                                                 url  \
0  https://lendingclub.com/browse/loanDetail.acti...   
1  https://lendingclub.com/browse/loanDetail.acti...   
2  https://lendingclub.com/browse/loanDetail.acti...   
3  https://lendingclub.com/browse/loanDetail.acti...   
4  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65            0   
1                  bike    309xx         GA   1.00            0   
2  real estate business    606xx         IL   8.72            0   
3              personel    917xx         CA  20.00            0   
4              Personal    972xx         OR  17.94            0   

  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal revol_util  \
0           Jan-85               1         3        0      13648     83.70%   
1           Apr-99               5         3        0       1687      9.40%   
2           Nov-01               2         2        0       2956     98.50%   
3           Feb-96               1        10        0       5598        21%   
4           Jan-96               0        15        0      27783     53.90%   

   total_acc  out_prncp  out_prncp_inv   total_pymnt  total_pymnt_inv  \
0          9       0.00           0.00   5863.155187          5833.84   
1          4       0.00           0.00   1008.710000          1008.71   
2         10       0.00           0.00   3005.666844          3005.67   
3         37       0.00           0.00  12231.890000         12231.89   
4         38     524.06         524.06   3513.330000          3513.33   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5000.00         863.16                0.00        0.00   
1           456.46         435.17                0.00      117.08   
2          2400.00         605.67                0.00        0.00   
3         10000.00        2214.92               16.97        0.00   
4          2475.94        1037.39                0.00        0.00   

   collection_recovery_fee last_pymnt_d  last_pymn

### Columns - issue_d

These is a date column. The month and year can be extracted from this column.

In [24]:
loan_data['issue_d'] = loan_data['issue_d'].apply(pd.to_datetime, format='%b-%y')
loan_data['issue_year'] = loan_data['issue_d'].dt.year
loan_data['issue_month'] = loan_data['issue_d'].dt.month
loan_data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0  1077501    1296599       5000         5000           4975.0   
1  1077430    1314167       2500         2500           2500.0   
2  1077175    1313524       2400         2400           2400.0   
3  1076863    1277178      10000        10000          10000.0   
4  1075358    1311748       3000         3000           3000.0   

   term_in_months  int_rate  installment grade sub_grade  \
0              36     10.65       162.87     B        B2   
1              60     15.27        59.83     C        C4   
2              36     15.96        84.33     C        C5   
3              36     13.49       339.31     C        C1   
4              60     12.69        67.79     B        B5   

                  emp_title  emp_length home_ownership  annual_inc  \
0                       NaN          10           RENT     24000.0   
1                     Ryder           0           RENT     30000.0   
2                       NaN          10           RENT     12252.0   
3       AIR RESOURCES BOARD          10           RENT     49200.0   
4  University Medical Group           1           RENT     80000.0   

  verification_status    issue_d  loan_status  \
0            Verified 2011-12-01   Fully Paid   
1     Source Verified 2011-12-01  Charged Off   
2        Not Verified 2011-12-01   Fully Paid   
3     Source Verified 2011-12-01   Fully Paid   
4     Source Verified 2011-12-01      Current   

                                                 url  \
0  https://lendingclub.com/browse/loanDetail.acti...   
1  https://lendingclub.com/browse/loanDetail.acti...   
2  https://lendingclub.com/browse/loanDetail.acti...   
3  https://lendingclub.com/browse/loanDetail.acti...   
4  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65            0   
1                  bike    309xx         GA   1.00            0   
2  real estate business    606xx         IL   8.72            0   
3              personel    917xx         CA  20.00            0   
4              Personal    972xx         OR  17.94            0   

  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal revol_util  \
0           Jan-85               1         3        0      13648     83.70%   
1           Apr-99               5         3        0       1687      9.40%   
2           Nov-01               2         2        0       2956     98.50%   
3           Feb-96               1        10        0       5598        21%   
4           Jan-96               0        15        0      27783     53.90%   

   total_acc  out_prncp  out_prncp_inv   total_pymnt  total_pymnt_inv  \
0          9       0.00           0.00   5863.155187          5833.84   
1          4       0.00           0.00   1008.710000          1008.71   
2         10       0.00           0.00   3005.666844          3005.67   
3         37       0.00           0.00  12231.890000         12231.89   
4         38     524.06         524.06   3513.330000          3513.33   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5000.00         863.16                0.00        0.00   
1           456.46         435.17                0.00      117.08   
2          2400.00         605.67                0.00        0.00   
3         10000.00        2214.92               16.97        0.00   
4          2475.94        1037.39                0.00        0.00   

   collection_recovery_fee last_

## Filter Data

### Deduplicate Rows

If there are any duplicate rows in the dataset, these can result in incorrect results during analysis and these rows can hence be dropped. In the current dataset, all rows are having an id column. We can check the duplicates in id field to check this.

In [25]:
loan_data['id'].duplicated().sum()

0

As per the check above, there are no duplicate rows. Hence nothing need to be dropped.

### Exclude loan_status = Current

We can filter out rows where loan_status is Current. If the status is current, we are not yet sure whether the borrower is going to default or not. So these rows won't help in analysis.

In [26]:
existing_row_count = len(loan_data)
loan_data = loan_data[loan_data['loan_status'] != 'Current']
current_row_count = len(loan_data)
print('Dropped %d rows' % (existing_row_count - current_row_count))

Dropped 1098 rows


### Exclude Irrelavant Columns

The columns available are:

In [27]:
loan_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37544 entries, 0 to 39716
Data columns (total 47 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   id                       37544 non-null  int64         
 1   member_id                37544 non-null  int64         
 2   loan_amnt                37544 non-null  int64         
 3   funded_amnt              37544 non-null  int64         
 4   funded_amnt_inv          37544 non-null  float64       
 5   term_in_months           37544 non-null  int64         
 6   int_rate                 37544 non-null  float64       
 7   installment              37544 non-null  float64       
 8   grade                    37544 non-null  object        
 9   sub_grade                37544 non-null  object        
 10  emp_title                36136 non-null  object        
 11  emp_length               37544 non-null  int64         
 12  home_ownership           37544 n

The columns that can be dropped are:
* id: Id is a unique value which identify a loan. No information can be derived from id.
* member_id: Unique value to identify a person. No information can be derived from member_id.
* funded_amnt: This is a value determined after the loan is approved. Hence its not useful to predict if the borrower will default.
* funded_amnt_inv: This is a value determined after the loan is approved. Hence its not useful to predict if the borrower will default.
* emp_title: This is a title of the person with regard to work. This has too many values to be for categorizing and there is no standard way to combine or compare two titles of people from different work streams.
* title: This is the loan title. This is again a description and it would be difficult to use this to find patterns.
* url: This is not a useful field as url only has an id field which is differing across rows. This id field cannot provide any relevant info.
* desc: This is a general description of the loan. This cannot be used for classifications. Purpose is a more standard and is similar to desc.
* zip_code: This field is masked. Hence cannot be used.
* Below attributes are affected by whether the loan is approved or not. Hence these cannot be relied on for analysis:
    * revol_bal
    * revol_util
    * out_prncp
    * out_prncp_inv
    * total_pymnt
    * total_pymnt_inv
    * total_rec_prncp
    * total_rec_int
    * total_rec_late_fee
    * collection_recovery_fee
    * last_pymnt_d
    * last_pymnt_amnt
    * last_credit_pull_d

In [28]:
columns_to_be_dropped = [
    'id',
    'member_id',
    'funded_amnt',
    'funded_amnt_inv',
    'emp_title',
    'title',
    'url',
    'desc',
    'zip_code',
    'revol_bal',
    'revol_util',
    'out_prncp',
    'out_prncp_inv',
    'total_rec_prncp',
    'total_rec_int',
    'total_rec_late_fee',
    'collection_recovery_fee',
    'last_pymnt_d',
    'last_pymnt_amnt',
    'last_credit_pull_d'
]
loan_data = loan_data.drop(columns=columns_to_be_dropped)
loan_data

loan_amnt  term_in_months  int_rate  installment grade sub_grade  \
0           5000              36     10.65       162.87     B        B2   
1           2500              60     15.27        59.83     C        C4   
2           2400              36     15.96        84.33     C        C5   
3          10000              36     13.49       339.31     C        C1   
5           5000              36      7.90       156.46     A        A4   
...          ...             ...       ...          ...   ...       ...   
39712       2500              36      8.07        78.42     A        A4   
39713       8500              36     10.28       275.38     C        C1   
39714       5000              36      8.07       156.84     A        A4   
39715       5000              36      7.43       155.38     A        A2   
39716       7500              36     13.75       255.43     E        E2   

       emp_length home_ownership  annual_inc verification_status    issue_d  \
0              10           RENT     24000.0            Verified 2011-12-01   
1               0           RENT     30000.0     Source Verified 2011-12-01   
2              10           RENT     12252.0        Not Verified 2011-12-01   
3              10           RENT     49200.0     Source Verified 2011-12-01   
5               3           RENT     36000.0     Source Verified 2011-12-01   
...           ...            ...         ...                 ...        ...   
39712           4       MORTGAGE    110000.0        Not Verified 2007-07-01   
39713           3           RENT     18000.0        Not Verified 2007-07-01   
39714           0       MORTGAGE    100000.0        Not Verified 2007-07-01   
39715           0       MORTGAGE    200000.0        Not Verified 2007-07-01   
39716           0            OWN     22000.0        Not Verified 2007-06-01   

       loan_status             purpose addr_state    dti  delinq_2yrs  \
0       Fully Paid         credit_card         AZ  27.65            0   
1      Charged Off                 car         GA   1.00            0   
2       Fully Paid      small_business         IL   8.72            0   
3       Fully Paid               other         CA  20.00            0   
5       Fully Paid             wedding         AZ  11.20            0   
...            ...                 ...        ...    ...          ...   
39712   Fully Paid    home_improvement         CO  11.33            0   
39713   Fully Paid         credit_card         NC   6.40            1   
39714   Fully Paid  debt_consolidation         MA   2.30            0   
39715   Fully Paid               other         MD   3.72            0   
39716   Fully Paid  debt_consolidation         MA  14.29            1   

      earliest_cr_line  inq_last_6mths  open_acc  pub_rec  total_acc  \
0               Jan-85               1         3        0          9   
1               Apr-99               5         3        0          4   
2               Nov-01               2         2        0         10   
3               Feb-96               1        10        0         37   
5               Nov-04               3         9        0         12   
...                ...             ...       ...      ...        ...   
39712           Nov-90               0        13        0         40   
39713           Dec-86               1         6        0          9   
39714           Oct-98               0        11        0         20   
39715           Nov-88               0        17        0         26   
39716           Oct-03               0         7        0          8   

        total_pymnt  total_pymnt_inv  recoveries  pub_rec_bankruptcies  \
0       5863.155187          5833.84        0.00                   0.0   
1       1008.710000          1008.71      117.08                   0.0   
2       3005.666844          3005.67        0.00                   0.0   
3      12231.890000         12231.89        0.00                   0.0   
5       5632.210000          5632.21        0

## Misc Corrections
Once the data is filtered, few more minor corrections can be done.

### Update Missing Values
Find and fix the missing values

In [34]:
loan_data.isnull().sum()

loan_amnt                 0
term_in_months            0
int_rate                  0
installment               0
grade                     0
sub_grade                 0
emp_length                0
home_ownership            0
annual_inc                0
verification_status       0
issue_d                   0
loan_status               0
purpose                   0
addr_state                0
dti                       0
delinq_2yrs               0
earliest_cr_line          0
inq_last_6mths            0
open_acc                  0
pub_rec                   0
total_acc                 0
total_pymnt               0
total_pymnt_inv           0
recoveries                0
pub_rec_bankruptcies    697
issue_year                0
issue_month               0
dtype: int64

We can see that only pub_rec_bankruptcies is having missing values. These could be because there are no bankruptcy record found for the borrower, meaning it would be zero. We can default the nulls as zero here.

In [40]:
loan_data['pub_rec_bankruptcies'] = loan_data['pub_rec_bankruptcies'].fillna(0).astype(int)

# Univariate Analysis

In [29]:
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

In [30]:
# Utility functions

# IQR
def drop_outliers(dataframe, column):
    perc75 = dataframe[column].quantile(0.75)
    perc25 = dataframe[column].quantile(0.25)
    iqr = perc75-perc25
    lower_bound = perc25 - iqr * 1.5
    upper_bound = perc75 + iqr * 1.5
    current_rows = len(dataframe)
    dataframe = dataframe[(dataframe[column] >= lower_bound) & (dataframe[column] <= upper_bound)]
    print('Dropped %s rows' % (current_rows - len(dataframe)))
    return dataframe    
    

In [31]:
## Remove Outliers
plt.figure(figsize=(20,10))
sns.boxplot(loan['loan_amnt'], orient='h')
drop_outliers()

NameError: name 'loan' is not defined

<Figure size 2000x1000 with 0 Axes>

In [ ]:
sns.boxplot(drop_outliers(loan, 'loan_amnt')['loan_amnt'], orient='h')